man/gd_alter_endianness.3.in
.\" gd_alter_endianness.3.  The gd_alter_endianness man page.
.\"
.\" @configure_input@
.\"
.\" Copyright (C) 2008, 2010, 2012, 2014, 2016 D. V. Wiebe
.\"
.\""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""""
.\"
.\" This file is part of the GetData project.
.\"
.\" Permission is granted to copy, distribute and/or modify this document
.\" under the terms of the GNU Free Documentation License, Version 1.2 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, with no Front-Cover Texts, and with no Back-Cover
.\" Texts.  A copy of the license is included in the `COPYING.DOC' file
.\" as part of this distribution.
.\"
.TH gd_alter_endianness 3 "22 November 2016" "Version 0.10.0" "GETDATA"
.SH NAME
gd_alter_endianness \(em modify the byte sex of fields in a dirfile
.SH SYNOPSIS
.B #include <getdata.h>
.HP
.nh
.ad l
.BI "int gd_alter_endianness(DIRFILE *" dirfile ", unsigned long"
.IB byte_sex ", int " fragment_index ", int " recode );
.hy
.ad n
.SH DESCRIPTION
The
.BR gd_alter_endianness ()
function sets the byte sex of the format specification fragment given by
.I fragment_index
to
.I byte_sex
in the dirfile(5) database specified by
.IR dirfile .
The byte sex of a fragment indicate the endianness of data stored in binary
files associated with
.B RAW
fields defined in the specified fragment.  The byte sex of a fragment containing
no
.B RAW
fields is ignored.

The
.I byte_sex
argument should be one of the following:
.TP
.BR 0 " (zero)"
Indicating that the byte sex should be the native endianness of the host,
whichever that may be.
.TP
.B GD_BIG_ENDIAN
Indicating that the byte sex should be big endian.
.TP
.B GD_LITTLE_ENDIAN
Indicating that the byte sex should be little endian.
.TP
.RB ( GD_BIG_ENDIAN " | " GD_LITTLE_ENDIAN )
Indicating that the byte sex should be the opposite of the native endianness of
the host, whichever that may be.
.PP
Furthermore, any of these may be bitwise or'd
with
.B GD_ARM_ENDIAN
or
.B GD_NOT_ARM_ENDIAN
indicating that the floating point data are stored in the ARM middle-endian
format.
.PP
In addition to being simply a valid fragment index,
.I fragment_index
may also be the special value
.BR GD_ALL_FRAGMENTS ,
which indicates that the byte sex of all fragments in the database should
be changed.

If the
.I recode
argument is non-zero, this call will byte swap the binary data of affected
.B RAW
fields to account for the change in byte sex.  If the encoding of the
fragment is endianness insensitive, or if the data type is only one byte in
size, no change is made.  The I/O pointer of all affected
.B RAW
fields is reset to the beginning-of-frame.

If
.I recode
is zero, affected binary files are left untouched.

.SH RETURN VALUE
Upon successful completion,
.BR gd_alter_endianness ()
returns zero.  On error, it returns a negative-valued error code.  Possible
error codes are:
.TP 8
.B GD_E_ACCMODE
The specified dirfile was opened read-only.
.TP
.B GD_E_ALLOC
The library was unable to allocate memory.
.TP
.B GD_E_ARGUMENT
The supplied
.I byte_sex
was invalid.
.TP
.B GD_E_BAD_DIRFILE
The supplied dirfile was invalid.
.TP
.B GD_E_BAD_INDEX
The supplied index was out of range.
.TP
.B GD_E_IO
An I/O error occurred while attempting to byte swap a binary file.
.TP
.B GD_E_PROTECTED
The metadata of the indicated format specification fragment was protected from
change, or the binary data of the fragment was protected from change and binary
file byte swapping was requested.
.TP
.B GD_E_UNCLEAN_DB
An error occurred while moving the byte-swapped file into place.  As a result,
the database may be in an unclean state.  See the
.B NOTES
section below for recovery instructions.  In this case, the dirfile will be
flagged as invalid, to prevent further database corruption.  It should be
immediately closed.
.TP
.B GD_E_UNKNOWN_ENCODING
The encoding scheme of the fragment is unknown.
.TP
.B GD_E_UNSUPPORTED
The encoding scheme of the fragment does not support binary file byte swapping.
.PP
The error code is also stored in the
.B DIRFILE
object and may be retrieved after this function returns by calling
.BR gd_error (3).
A descriptive error string for the error may be obtained by calling
.BR gd_error_string (3).
.SH NOTES
A binary file byte swap occurs out-of-place.  As a result, sufficient space
must be present on the filesystem for the binary files of all
.B RAW
fields in the fragment both before and after translation.  If all fragments
are updated by specifying
.BR GD_ALL_FRAGMENTS ,
the byte swapping occurs one fragment at a time.

An error code of
.B GD_E_UNCLEAN_DB
indicates a system error occurred while moving the byte-swapped binary data into
place or when deleting the old data.  If this happens, the database may be left
in an unclean state.  The caller should check the filesystem directly to
ascertain the state of the dirfile data before continuing.  For recovery
instructions, see the file
@absolute_docdir@/unclean_database_recovery.txt.
.SH SEE ALSO
.BR gd_open (3),
.BR gd_error (3),
.BR gd_error_string (3),
.BR gd_endianness (3),
.BR dirfile (5),
.BR dirfile-format (5)


NEWS
New in version 0.10.0a:

  Dirfile Changes:

  * Dirfile Standards Version 10 has been released.  It adds three new
    field types: SARRAY, which is an array of STRING scalars (like a
    CARRAY, but with STRINGs instead of CONSTs), and INDIR and SINDIR,
    which are vector fields provide indexed look-ups from CARRAY (INDIR) or
    SARRAY (SINDIR) scalar arrays.  It also adds field code namespaces,
    which can be specified with the new /NAMESPACE directive, or else as
    part of an /INCLUDE directive.

  * Some notes about namespaces:

    - Namespaces are separated from field names by a dot (.):
  
          namespace.name
        
      and can be nested arbitrarily deep, separating namespaces with
      intermediate dots:

          namespace.subspace.subsubspace.name

      Namespaces were created to provide an alternative to the prefix and
      suffix added to the /INCLUDE driective in Standards Version 9, which
      have some unfortunate side-effects due to their modifying field names
      directly.  Because namespaces nest and are syntactically separate
      from field names, they do a better job of encapsulation.

    - If the namespace of a field is null (""), which is the default, then
      the dot separing the namespace from the field name may be omitted.

    - An /INCLUDE statement can specify a namespace which becomes the
      included fragment's "root namespace".  The root namespace of the base
      (top-level) format file is always the nullspace (""), and cannot be
      changed.

    - In additon to root namespaces, there is also a "current namespace".
      At the top of a fragment, the current namespace is set to the root
      namespace.  A /NAMESPACE directive can be used to change the current
      namespace to a subspace under the root namespace.  That is, if the
      root namespace is "root", then the directive:

          /NAMESPACE subspace

       changes the current namespace to "root.subspace" regardless of what
       the current namespace was before.  To change the current namespace
       back to the root namespace, use the null token ("") with the
       /NAMESPACE directive:

          /NAMESPACE ""

    - If no namespace is specified in an /INCLUDE line, then the current
      namespace becomes the root namespace of the included fragment.

    - Subnamespaces under the root namespace may also be specified directly
      in the name part of a field specification.  As a result, it is never
      necessary to use the /NAMESPACE directive.

    - Every field code and name in a fragment implicitly gains either the
      fragment's root namespace or the current namespace.  If the field
      code starts with an initial dot, then the root namespace is prepended
      to it, otherwise the current namespace is prepended to it.  Because
      the current namespace is always a subspace of the root namespace,
      this means that metadata in a given fragment is never able to access
      other fields outside its own root namespace.

    - The exception to the above is the implicit INDEX vector, which
      ignores all namespaces attached to it, either implicitly, through the
      current namespace, or explicitly when specified in the metadata.
      This contrasts with the behaviour of the INDEX field in the presence
      of affixes, where it can appears or disappears based on the effects
      of the affixes creating or modifying the specific field name "INDEX".

  * Because syntactially the dot (.) now performs two functions, namely
    both separating namespaces from each other and field names, but also
    separating a field code from a representation suffix, there exists
    ambiguity in the syntax.  To resolve the ambiguity, a new representa-
    tion suffix, .z has been added which does nothing.  As an example, the
    field code:

      name.r

    is interpreted as the real part of the field named "name", asuming such
    a field exists.  To indicate the field named "r" in the namespace
    "name", the field code:

      name.r.z

    must be used.  Note that this ambiguity only exists in a Dirfile where
    both "name" and "name.r" are valid field names.  If the field named
    "name" doesn't exist, then the first field code, "name.r" is unambig-
    uously interpreted as the field "r" in the namespace "name".

  * A note on the SINDIR field: Unlike every other vector field, this
    field produces character string data at the sample rate of it's input
    vector, which may be surprising, in certain instances.

  * GetData has supported FLAC compression since 0.9.0.  Standards Version
    10 now adds "flac" to the list of pre-defined encodings.

  Library Changes:

  * The function gd_array_len() no longer silently ignores a representation
    suffix in the field_code provided.  In most cases, passing a
    representation suffix will now cause this function to fail with a
    GD_E_BAD_CODE error.

  * A couple of unnecessary malloc's have been removed from the field
    code search, reducing the chance of encountering a GD_E_ALLOC error.
    Notably the functions gd_array_len(), gd_bof(), gd_entry_type(),
    gd_fragment_index(), and gd_spf() will no longer produce this error at
    all.

  * BUG FIX: Writes to ASCII (text) encoded files weren't properly updating
    the file's I/O position, leading to subsequent reads and writes
    occurring in the wrong place.

  * BUG FIX: Several bugs in the I/O positioning performed before reads and
    writes to compressed data have been fixed.  Previously, reads and
    writes could occur in the wrong place in the data stream.  Reported by
    S. J. Benton.

  * BUG FIX: Trying to open a non-existent, gzip-encoded data file now
    reports the correct error (GD_E_IO; "No such file or directory"),
    instead of segfaulting.  Reported by Matthew Petroff.

  * BUG FIX: A segfault encountered when closing very large compressed
    files after writing to them has been fixed.


bindings/matlab/gd_alter_endianness.c
/* Copyright (C) 2013 D. V. Wiebe
 *
 ***************************************************************************
 *
 * This file is part of the GetData project.
 *
 * GetData is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 *
 * GetData is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with GetData; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
 */
#include "gd_matlab.h"

/*
 % GD_ALTER_ENDIANNESS  Modify the byte sex of a fragment
 %
 %   GD_ALTER_ENDIANNESS(DIRFILE,ENDIANNESS,FRAGMENT[,RECODE])
 %             sets the byte sex of the metadata fragment indexed by FRAGMENT
 %             in the dirfile DIRFILE to ENDIANNESS, which should be one of:
 %             numeric zero, indicating the native type of the platform;
 %             GD.BIG_ENDIAN, or GD.LITTLE_ENDIAN, optionally bitwise or'd with
 %             either GD.ARM_ENDIAN or GD.NOT_ARM_ENDIAN.  See GETDATA_CONSTANTS
 %             for these constants.  If RECODE is given and non-zero, binary
 %             files affected by the byte sex change will be converted.
 %             Otherwise, the binary files are not modified.
 %
 %   The DIRFILE object should have previously been created with GD_OPEN.
 %
 %   See the documentation on the C API function gd_alter_endianness(3) in
 %   section 3 of the UNIX manual for more details.
 %
 %   See also GD_ENDIANNESS, GD_OPEN, GETDATA_CONSTANTS
 */

void mexFunction(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
{
  DIRFILE *D;
  void *data;
  int i, r = 0;
  unsigned long e;

  GDMX_NO_LHS;
  GDMX_CHECK_RHS2(3,4);

  D = gdmx_to_dirfile(prhs[0]);
  e = gdmx_to_ulong(prhs, 1);
  i = gdmx_to_int(prhs, 2);
  if (nrhs > 3)
    r = gdmx_to_int(prhs, 3);

  gd_alter_endianness(D, e, i, r);

  gdmx_err(D, 0);
}
